packages/firestore/src/core/sync_engine.ts

```diff
@@ -367,30 +368,41 @@ export class SyncEngine implements RemoteSyncer, SharedClientStateSyncer {
    */
   async runTransaction<T>(
     updateFunction: (transaction: Transaction) => Promise<T>,
+    backoff: ExponentialBackoff,
     retries = 5
   ): Promise<T> {
     assert(retries >= 0, 'Got negative number of retries for transaction.');
-    const transaction = this.remoteStore.createTransaction();
-    const userPromise = updateFunction(transaction);
-    if (
-      isNullOrUndefined(userPromise) ||
-      !userPromise.catch ||
-      !userPromise.then
-    ) {
-      return Promise.reject<T>(
-        Error('Transaction callback must return a Promise')
-      );
-    }
-    try {
-      const result = await userPromise;
-      await transaction.commit();
-      return result;
-    } catch (error) {
-      if (retries > 0 && this.isRetryableTransactionError(error)) {
-        return this.runTransaction(updateFunction, retries - 1);
+    let transactionError;
+    for (let i = 0; i < retries; i++) {
+      const backoffDeferred = new Deferred<void>();
+      backoff.backoffAndRun(async () => {
+        backoffDeferred.resolve();
+      });
+
+      try {
+        await backoffDeferred.promise;
```

Comment:
This is a neat use of async/await to do the retries in a for-loop, but since we don't have async/await on the other platforms, it's probably not going to be very portable. :-/  I'd suggest implementing transaction retries on Android/iOS and then revisiting this approach to try to keep the platforms in sync.

If the non-async/await approach ends up being sufficiently heinous, then perhaps it's worth still doing it on JavaScript, but I'm hoping it's not too bad.

Reply:
I attempted a draft on Android. I'm not certain that it works yet since my integration tests won't run, but can you take a look to see if it's in the right spiritual direction? ([link](https://github.com/firebase/firebase-android-sdk/compare/bc/tx-backoff?expand=1))